libyul/backends/evm/StackHelpers.h

```diff
@@ -0,0 +1,363 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/backends/evm/ControlFlowGraph.h>
+#include <libyul/Exceptions.h>
+
+#include <libsolutil/Visitor.h>
+
+#include <range/v3/algorithm/all_of.hpp>
+#include <range/v3/algorithm/any_of.hpp>
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/iota.hpp>
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/take.hpp>
+
+namespace solidity::yul
+{
+
+inline std::string stackSlotToString(StackSlot const& _slot)
+{
+	return std::visit(util::GenericVisitor{
+		[](FunctionCallReturnLabelSlot const& _ret) -> std::string { return "RET[" + _ret.call.get().functionName.name.str() + "]"; },
+		[](FunctionReturnLabelSlot const&) -> std::string { return "RET"; },
+		[](VariableSlot const& _var) { return _var.variable.get().name.str(); },
+		[](LiteralSlot const& _lit) { return util::toCompactHexWithPrefix(_lit.value); },
+		[](TemporarySlot const& _tmp) -> std::string { return "TMP[" + _tmp.call.get().functionName.name.str() + ", " + std::to_string(_tmp.index) + "]"; },
+		[](JunkSlot const&) -> std::string { return "JUNK"; }
+	}, _slot);
+}
+
+inline std::string stackToString(Stack const& _stack)
+{
+	std::string result("[ ");
+	for (auto const& slot: _stack)
+		result += stackSlotToString(slot) + ' ';
+	result += ']';
+	return result;
+}
+
+
+// Abstraction of stack shuffling operations. Can be defined as actual concept once we switch to C++20.
+/*
+template<typename ShuffleOperations>
+concept ShuffleOperationConcept = requires(ShuffleOperations ops, size_t sourceOffset, size_t targetOffset, size_t depth) {
+	// Returns true, iff the current slot at sourceOffset in source layout is a suitable slot at targetOffset.
```

Comment:
"source layout" is rather "current layout" while "target layout" is always fixed, is it?

Reply:
What does "suitable" mean?

Reply:
> "source layout" is rather "current layout" while "target layout" is always fixed, is it?

yep

Reply:
> What does "suitable" mean?

Whatever the user considers "suitable". Usually that the slots are either identical or the target slot can be arbitrary. But in "createIdealLayout" also a symbolic slot (resp. "PreviousSlot") is compatible with any other slot (except those in the operation output, since they need fixed positions).